Source/NSSpeechRecognizer.m

```diff
@@ -23,8 +23,176 @@
 */
 
 #import <AppKit/NSSpeechRecognizer.h>
+#import <Foundation/NSDistantObject.h>
+#import <Foundation/NSString.h>
+#import <Foundation/NSDictionary.h>
+#import <Foundation/NSArray.h>
+#import <Foundation/NSThread.h>
+#import <Foundation/NSError.h>
+#import <Foundation/NSConnection.h>
+#import <Foundation/NSDistributedNotificationCenter.h>
+#import "GSFastEnumeration.h"
+#import "AppKit/NSWorkspace.h"
 
-@implementation NSSpeechRecognizer
+id   _speechRecognitionServer = nil;
+BOOL _serverLaunchTested = NO;
+
+#define SPEECH_RECOGNITION_SERVER @"GSSpeechRecognitionServer"
 
+@interface NSObject (GSSpeechRecognitionServer)
+- (NSSpeechRecognizer *) newRecognizer;
 @end
 
+@implementation NSSpeechRecognizer
+
++ (void) initialize
+{
+  if (self == [NSSpeechRecognizer class])
+    {
+      _speechRecognitionServer = [NSConnection
+                                   rootProxyForConnectionWithRegisteredName: SPEECH_RECOGNITION_SERVER
+                                                                       host: nil];
+      RETAIN(_speechRecognitionServer);
+      if (nil == _speechRecognitionServer)
+        {
+          NSWorkspace *ws = [NSWorkspace sharedWorkspace];
+          [ws launchApplication: SPEECH_RECOGNITION_SERVER
+                       showIcon: NO
+                     autolaunch: NO];
+        }
+      else
+        {
+          NSLog(@"Server found in +initialize");
+        }
+    }
+}
+
+- (void) processNotification: (NSNotification *)note
+{
+  NSString *word = (NSString *)[note object];
+
+  word = [word lowercaseString];
+  FOR_IN(NSString*, obj, _commands)
+    {
+      if ([[obj lowercaseString] isEqualToString: word])
+        {
+          [_delegate speechRecognizer: self
+                  didRecognizeCommand: word];
+        }
+    }
+  END_FOR_IN(_commands);
+}
+
+// Initialize
+- (instancetype) init
+{
+  self = [super init];
+  if (self != nil)
+    {
+      [[NSDistributedNotificationCenter defaultCenter]
```

Comment:
As I stated further down, I am really confused where which part of the code of this class will get executed. If this code here happens on the server, than we wouldn't need a distributed notification. Could you please clear up your view of the general working of this code?

Reply:
Sure...  one sec....

1)  When the application is loaded the first thing that happens is that +[NSSpeechRecognizer initialize] is called.  This method attempts to connect with the server (this covers the case where the server is already running from a previous invocation). 

2) if the server isnt running, it is started using workspace as the server is registered as an application. (like the synthesizer)

3) when -[NSSpeechRecognizer allocWithZone] is called it will try to connect to the server if it doesn't currently exist and keep doing so over 5 seconds until it gets the connection.  Once it has the connection it returns the concrete implementation provided by the GSSpeecRecognitionServer which is GSSpeechRecognizer.  

4) +[GSSpeechRecognizer initialize] then registers the "connectionDied" method to respond to NSConnectionDidDIeNotification in case the NSConnection dies.   It also initializes the shared server.   The shared server is a subclass of GSSpeechRecognitionServer.

5) In GSSpeechRecognitionServer init, it instantiates the GSSpeechRecognitionEngine which returns the concrete instance of PocketsphinxSpeechRecognitionEngine.

6)  At this point, when the method  startListening or stopListening are invoked it is done so on the GSSpeechRecognizer instance which was return and invokes _engine startListening or stopListening.   This method calls the code needed to start and stop the thread which contains the code which is listening to the mic input.

As discussed above.  This code is based on the NSSpeechSynthesizer "architecture".   I have no issues simplifying it.   As it is it is crashing when calling startListening and I am not sure why.

Reply:
This should now be resolved.  I have simplified the architecture.